### PR TITLE
Improve Functor instance for maps

### PR DESCRIPTION
**Pull request by sergv**

### CODE REVIEW

**Comment by treeowl** [comment]
Could you give some evidence that this is actually better? That's not exactly obvious, and the performance considerations of minor changes like this can be hard to guess.

**Comment by chessai** [comment]
i agree with David here. could this be benchmarked?

**Comment by sergv** [comment]
Okay, I'll try to come up with something.

**Comment by sergv** [comment]
**TL;DR** there's no significant difference, and even a slight slowdown with newer version, so PR can probably be closed at this point. Although perhaps removing arguments from `fmap` is a strict improvement since it would only improve its inlining and nothing more.

**Longer version:** I did some benchmarks with `Data.Map.Internal` to try to find out whether defining `<$` via local helper vs current recursive version that passes argument on each call would make any difference. It turned out that the local helper version was always slightly behind current version for most non-trivial map sizes (10^k, k=1..6). That was surprising but it turned out that in C-- version defined via local helper does basically the same thing but contains a few extra instructions on each recursive call. Unfortunately, I haven't traced what causes those instructions to appear or what they do.

After this I came up with a simpler benchmark to try to illustrate the issue. I defined simple recursive tree data type, which should be similar to the `Map`: `data Tree a = Tip | Bin a (Tree a) (Tree a)` and a spine-strict variant `data Tree' a = Tip' | Bin' a (Tree' a) (Tree' a)`. Then each of them got a version of `<$` that recurses directly or uses a local helper, `mapConstRec` and `mapConstHepler'` respectively for `Tree` and same names with a `'` for `Tree'`. My conclusion is that version defined via recursive helper *is* faster that the recusive one for *lazy* recursive data types. For spine-strict data types, of which `Map` is an instance of, that is not the case - passing argument each time doesn't seem to matter for the runtime.

Simpler benchmark results, perfect trees of size `(2^height - 1)`:
```
height "1"/mapConstRec                   mean 13.98 ns  ( +- 84.33 ps  )
height "1"/mapConstHelper                mean 14.59 ns  ( +- 101.2 ps  )
height "1"/mapConstRec'                  mean 11.60 ns  ( +- 32.43 ps  )
height "1"/mapConstHelper'               mean 12.30 ns  ( +- 31.98 ps  )
height "2"/mapConstRec                   mean 32.36 ns  ( +- 277.2 ps  )
height "2"/mapConstHelper                mean 31.66 ns  ( +- 326.6 ps  )
height "2"/mapConstRec'                  mean 25.01 ns  ( +- 168.2 ps  )
height "2"/mapConstHelper'               mean 25.18 ns  ( +- 102.6 ps  )
height "3"/mapConstRec                   mean 76.53 ns  ( +- 696.1 ps  )
height "3"/mapConstHelper                mean 74.36 ns  ( +- 589.1 ps  )
height "3"/mapConstRec'                  mean 50.45 ns  ( +- 287.4 ps  )
height "3"/mapConstHelper'               mean 48.35 ns  ( +- 148.0 ps  )
height "4"/mapConstRec                   mean 142.2 ns  ( +- 999.4 ps  )
height "4"/mapConstHelper                mean 136.2 ns  ( +- 1.746 ns  )
height "4"/mapConstRec'                  mean 122.1 ns  ( +- 558.1 ps  )
height "4"/mapConstHelper'               mean 116.0 ns  ( +- 1.005 ns  )
height "5"/mapConstRec                   mean 298.8 ns  ( +- 2.537 ns  )
height "5"/mapConstHelper                mean 284.2 ns  ( +- 1.121 ns  )
height "5"/mapConstRec'                  mean 237.0 ns  ( +- 1.006 ns  )
height "5"/mapConstHelper'               mean 225.4 ns  ( +- 1.040 ns  )
height "6"/mapConstRec                   mean 604.6 ns  ( +- 7.216 ns  )
height "6"/mapConstHelper                mean 579.4 ns  ( +- 3.672 ns  )
height "6"/mapConstRec'                  mean 488.6 ns  ( +- 2.387 ns  )
height "6"/mapConstHelper'               mean 459.1 ns  ( +- 4.594 ns  )
height "7"/mapConstRec                   mean 1.261 μs  ( +- 9.875 ns  )
height "7"/mapConstHelper                mean 1.158 μs  ( +- 8.644 ns  )
height "7"/mapConstRec'                  mean 994.0 ns  ( +- 2.728 ns  )
height "7"/mapConstHelper'               mean 927.4 ns  ( +- 5.380 ns  )
height "8"/mapConstRec                   mean 2.561 μs  ( +- 31.80 ns  )
height "8"/mapConstHelper                mean 2.413 μs  ( +- 32.35 ns  )
height "8"/mapConstRec'                  mean 2.046 μs  ( +- 10.36 ns  )
height "8"/mapConstHelper'               mean 1.920 μs  ( +- 11.18 ns  )
height "9"/mapConstRec                   mean 5.168 μs  ( +- 77.64 ns  )
height "9"/mapConstHelper                mean 4.845 μs  ( +- 33.73 ns  )
height "9"/mapConstRec'                  mean 4.281 μs  ( +- 22.64 ns  )
height "9"/mapConstHelper'               mean 4.153 μs  ( +- 20.14 ns  )
height "10"/mapConstRec                  mean 10.45 μs  ( +- 154.3 ns  )
height "10"/mapConstHelper               mean 9.780 μs  ( +- 46.50 ns  )
height "10"/mapConstRec'                 mean 9.093 μs  ( +- 55.05 ns  )
height "10"/mapConstHelper'              mean 8.802 μs  ( +- 143.2 ns  )
height "11"/mapConstRec                  mean 21.08 μs  ( +- 292.3 ns  )
height "11"/mapConstHelper               mean 19.77 μs  ( +- 80.93 ns  )
height "11"/mapConstRec'                 mean 19.25 μs  ( +- 95.48 ns  )
height "11"/mapConstHelper'              mean 18.65 μs  ( +- 244.3 ns  )
height "12"/mapConstRec                  mean 42.01 μs  ( +- 671.9 ns  )
height "12"/mapConstHelper               mean 39.78 μs  ( +- 433.8 ns  )
height "12"/mapConstRec'                 mean 42.80 μs  ( +- 226.7 ns  )
height "12"/mapConstHelper'              mean 41.78 μs  ( +- 259.1 ns  )
height "13"/mapConstRec                  mean 84.41 μs  ( +- 1.299 μs  )
height "13"/mapConstHelper               mean 79.25 μs  ( +- 893.8 ns  )
height "13"/mapConstRec'                 mean 94.35 μs  ( +- 1.363 μs  )
height "13"/mapConstHelper'              mean 91.85 μs  ( +- 1.297 μs  )
height "14"/mapConstRec                  mean 156.6 μs  ( +- 2.300 μs  )
height "14"/mapConstHelper               mean 148.4 μs  ( +- 2.272 μs  )
height "14"/mapConstRec'                 mean 238.4 μs  ( +- 3.507 μs  )
height "14"/mapConstHelper'              mean 232.5 μs  ( +- 1.004 μs  )
height "15"/mapConstRec                  mean 442.1 μs  ( +- 8.699 μs  )
height "15"/mapConstHelper               mean 355.9 μs  ( +- 6.225 μs  )
height "15"/mapConstRec'                 mean 676.7 μs  ( +- 9.227 μs  )
height "15"/mapConstHelper'              mean 666.8 μs  ( +- 8.961 μs  )
height "16"/mapConstRec                  mean 1.291 ms  ( +- 26.08 μs  )
height "16"/mapConstHelper               mean 1.122 ms  ( +- 15.55 μs  )
height "16"/mapConstRec'                 mean 1.778 ms  ( +- 21.77 μs  )
height "16"/mapConstHelper'              mean 1.767 ms  ( +- 28.97 μs  )
height "17"/mapConstRec                  mean 3.031 ms  ( +- 26.82 μs  )
height "17"/mapConstHelper               mean 2.842 ms  ( +- 56.40 μs  )
height "17"/mapConstRec'                 mean 4.047 ms  ( +- 45.75 μs  )
height "17"/mapConstHelper'              mean 4.030 ms  ( +- 34.48 μs  )
height "18"/mapConstRec                  mean 6.720 ms  ( +- 628.5 μs  )
height "18"/mapConstHelper               mean 6.439 ms  ( +- 596.0 μs  )
height "18"/mapConstRec'                 mean 8.749 ms  ( +- 597.1 μs  )
height "18"/mapConstHelper'              mean 8.766 ms  ( +- 602.9 μs  )
height "19"/mapConstRec                  mean 16.39 ms  ( +- 2.343 ms  )
height "19"/mapConstHelper               mean 15.71 ms  ( +- 2.189 ms  )
height "19"/mapConstRec'                 mean 20.24 ms  ( +- 2.365 ms  )
height "19"/mapConstHelper'              mean 20.24 ms  ( +- 2.243 ms  )
height "20"/mapConstRec                  mean 36.42 ms  ( +- 3.788 ms  )
height "20"/mapConstHelper               mean 35.37 ms  ( +- 3.851 ms  )
height "20"/mapConstRec'                 mean 47.86 ms  ( +- 3.811 ms  )
height "20"/mapConstHelper'              mean 47.80 ms  ( +- 3.901 ms  )
```

Mostly self-contained benchmark source if you want to see for yourself:
```haskell
{-# LANGUAGE BangPatterns        #-}
{-# LANGUAGE DeriveGeneric       #-}
{-# LANGUAGE LambdaCase          #-}
{-# LANGUAGE ScopedTypeVariables #-}

{-# OPTIONS_GHC -O2 -ddump-simpl -dsuppress-uniques -dsuppress-idinfo -dsuppress-module-prefixes -dsuppress-type-applications -dsuppress-coercions -dppr-cols200 -dsuppress-type-signatures -ddump-to-file #-}

module Tree (main) where

import Control.DeepSeq
import Control.Exception
import GHC.Generics
import Gauge


data Tree a = Tip | Bin a (Tree a) (Tree a)
  deriving (Generic)

instance NFData a => NFData (Tree a)

mapConstRec :: b -> Tree a -> Tree b
mapConstRec b = \case
  Tip -> Tip
  Bin _ l r -> Bin b (mapConstRec b l) (mapConstRec b r)

mapConstHelper :: b -> Tree a -> Tree b
mapConstHelper b = go
  where
    go = \case
      Tip -> Tip
      Bin _ l r -> Bin b (go l) (go r)

perfectTree :: Int -> Tree Int
perfectTree = go 0
  where
    go n = \case
      0 -> Tip
      k -> Bin n (go (2 * n) k') (go (2 * n + 1) k')
        where
          !k' = k - 1

data Tree' a = Tip' | Bin' a !(Tree' a) !(Tree' a)
  deriving (Generic)

instance NFData a => NFData (Tree' a)

mapConstRec' :: b -> Tree' a -> Tree' b
mapConstRec' b = \case
  Tip' -> Tip'
  Bin' _ l r -> Bin' b (mapConstRec' b l) (mapConstRec' b r)

mapConstHelper' :: b -> Tree' a -> Tree' b
mapConstHelper' b = go
  where
    go = \case
      Tip' -> Tip'
      Bin' _ l r -> Bin' b (go l) (go r)

perfectTree' :: Int -> Tree' Int
perfectTree' = go 0
  where
    go n = \case
      0 -> Tip'
      k -> Bin' n (go (2 * n) k') (go (2 * n + 1) k')
        where
          !k' = k - 1

main :: IO ()
main = do
  let config = Gauge.defaultConfig

  let ts =
        [ (show k, perfectTree k, perfectTree' k)
        | k <- [1..20]
        ]

  evaluate $ rnf ts

  defaultMainWith config
    [ bgroup ("height " ++ show height)
        [ bench "mapConstRec"     $ nf (mapConstRec (1 :: Int)) t
        , bench "mapConstHelper"  $ nf (mapConstHelper (1 :: Int)) t
        , bench "mapConstRec'"    $ nf (mapConstRec' (1 :: Int)) t'
        , bench "mapConstHelper'" $ nf (mapConstHelper' (1 :: Int)) t'
        ]
    | (height, t, t') <- ts
    ]
```

**Comment by sjakobi** [comment]
> **TL;DR** there's no significant difference, and even a slight slowdown with newer version, so PR can probably be closed at this point. Although perhaps removing arguments from `fmap` is a strict improvement since it would only improve its inlining and nothing more.

Based on this summary I would indeed suggest to close the PR.

Many thanks for investigating the issue though, @sergv! :)

Thoughts, @treeowl?
